src/fastapibootcamp/handlers/external.py

```diff
@@ -317,3 +330,135 @@ async def post_log_demo(
+#    in exceptions.py)
+# 4. Raise these exceptions in your application code
+# 5. Optionally, add SlackRouteErrorHandler to the APIRouter in your
+#    handlers/endpoints module to automatically report any uncaught exception
+#    to Slack.
+#
+# If your error is caused by user input, you typically don't want to report
+# is to Slack, but instead to the user. In the astroplan application we'll
+# explore the ClientRequestError exception for this purpose in
+# handlers/astroplan/endpoints.py's get_observer function.
+#
+# Try it out:
+#   http post :8000/fastapi-bootcamp/error-demo custom_error:=true
+
+
+class ErrorRequestModel(BaseModel):
+    """Request model for the error POST endpoint."""
+
+    custom_error: bool = Field(
+        True, title="If true, raise the custom SlackException."
+    )
+
+
+@external_router.post(
+    "/error-demo", summary="Raise an internal service exception."
+)
+async def post_error_demo(data: ErrorRequestModel) -> JSONResponse:
+    """Raise an internal service exception."""
+    if data.custom_error:
+        raise DemoInternalError(
+            "A custom error occurred.", custom_data="Hello error!"
+        )
+
+    raise RuntimeError("A generic error occurred.")
+
+
+# =============================================================================
+# Lesson 6: Custom dependencies
+#
+# FastAPI dependencies are a way to add reusable code to your path operation
+# that's aware of the current request. Safir provides several dependencies,
+# see https://safir.lsst.io/api.html#module-safir.dependencies.arq etc.
+#
+# - arq_dependency provides a client to the Arq distributed job queue
+# - db_session_dependency provides a SQLAlchemy session
+# - auth_delegated_token_dependency provides a delegated token
+# - auth_dependency provides info about the current user
+# - auth_logger_dependency provides a logger with user info bound
+# - http_client_dependency provides an HTTPX async client
+# - logger_dependency provides a structlog logger (see Lesson 4)
+#
+# Besides these, you can create your own dependencies. In the astroplan
+# application we'll explore the request context dependency pattern.
+#
+# There are two types of dependencies you'll develop: functional dependencies
+# and class-based dependencies. These class-based dependencies can hold
+# persistent state that's reused across multiple requests.
+#
+# Try it out:
+#   http get :8000/fastapi-bootcamp/dependency-demo X-Custom-Header:foo
+
+# See src/fastapibootcamp/dependencies/demopersistentdependency.py for the
+# dependency that holds a persistent value. Below is a functional dependency:
+
+
+async def custom_header_dependency(request: Request) -> str | None:
+    """Provide the value of the X-Custom-Header from the request.
+
+    This shows how you can access the request object, query parameters, and
+    even other depdenencies in a dependency function's arguments.
+    """
+    return request.headers["X-Custom-Header"]
```

Comment:
I'm a tiny bit leery of this example just because FastAPI has this dependency built-in as:
```
x_custom_header: Annotated[str | None, Header()] = None
```
Maybe make it slightly more complicated, such as decoding a JSON blob that's stored in the header or something else that FastAPI can't do itself out of the box? Or maybe all you need is a comment noting that you wouldn't use a dependency just to retrieve a header.

Reply:
Maybe I can implement a reusable pattern for getting lists from query parameters like combining order, limit, and offset or cursor into a pydantic model?

Reply:
Ooo, yes, that's a great idea!